scene/2d/gpu_particles_2d.cpp

```diff
@@ -32,16 +32,20 @@
 
 #include "core/core_string_names.h"
 #include "scene/resources/particle_process_material.h"
+#include "scene/scene_string_names.h"
 
 #ifdef TOOLS_ENABLED
 #include "core/config/engine.h"
 #endif
 
 void GPUParticles2D::set_emitting(bool p_emitting) {
+	// Do not return even if `p_emitting == emitting` because `emitting` is just an approximation.
+
+	emitting = p_emitting;
 	RS::get_singleton()->particles_set_emitting(particles, p_emitting);
 
 	if (p_emitting && one_shot) {
-		set_process_internal(true);
+		restart();
```

Comment:
This introduces a significant change in behaviour. Is it necessary for this feature to work? 

Notably, it makes the behaviour of ``set_emitting()`` differ between CPUParticles and GPUParticles. So if this change is necessary, it needs to be made in both forms of particles

Reply:
The usage of `set_emitting` for one shot particle systems feels very unintended in my opinion. The resetting of the property and the time it takes till it can be done again feels wrong. It does not respect explosivness and I thought restarting would make sense. Moreover it saved some code copying for the timer reset. But you are right that this is a change of behaviour that shouldn't just be introduced.

With that said it seems nearly impossible to support the signal when using `set_emitting` without restarting. Setting the `emitting` property throws of the time prediction and in contrast to CPUParticles we can't just access the particles. So the other option I see is to only support the signal when `restart` is used. When using `one_shot` this method is the best option anyway. As long as it is documented, that usage of `emitting` can throw of the signal this seems ok to me.

For consistency the CPU particles could be changed as well to only support the signal with `restart`. But since the usage of emitting works fine there this discrepancy would be fine IMHO.

Reply:
I think the ideal solution would be to maintain the same behaviour as with CPUParticles in this PR (i.e. not reset in `set_emitting()` and then make another PR that adds the reset to `set_emitting()` for both GPUParticles and CPUParticles. I know the behaviour of setting the ``emitting`` property with `one_shot` is a common pain point for users, but I don't think this is the right place to address that pain point. especially since it breaks compatibility and creates a difference in behaviour between CPUParticles and GPUParticles. 

From my understanding, the problem with not restarting in ``set_emitting()`` for the signal is that the signal will fire when the particles _should_ have finished anyway right? even though they technically never finished. If that is the case, then couldn't we check if ``!p_emitting && one_shot && emitting`` and then in that case set a flag to cancel the signal?

Reply:
Actually since internal process is turned of when `!p_emitting` the problems would only occur if you were to turn off emitting and then turn it on again but at this point I have the feeling im overthinking the thing a little bit.

We should just reset the timer everytime emitting is turned on but not reset the gpu particles. The particle behaviour stays the same and the signal will always fire after the particles finished (though with some delay).

Reply:
That is correct. But at least it does not break existing behaviour. I just see no way to make an accurate estimate of remaining time when emitting is set to true without syncing the server. So either we say it will be late or we scrap the signal completly if restart is not used.

Reply:
Well, right now if emitting is set to false and then true before the full cycle has ended, it essentially just gets ignored. So I would do the same on the node side and just ignore it. 

IMO with the signal, we will already be a huge step up over the behaviour in master. Previously users would often set ``emitting`` to true while the cycle was going in the hopes that the cycle would continue or restart. Since that didn't work, they had to guess when the particle system finished so they could set emitting to true to start it again. With this PR, we can tell them just to wait for the signal before setting ``emitting`` to true. Alternatively, they can call ``restart()`` to restart immediately.